### PR TITLE
Add more dns attack support (#41)

### PR DESCRIPTION
**Pull request by sunny-xiao**

### CODE REVIEW

File: pkg/server/chaosd/network.go [hunk 1]
```diff
@@ -194,13 +205,65 @@ func (s *Server) applyTC(attack *core.NetworkCommand, ipset string, uid string)
...
+	reIp, err := regexp.Compile(`^(\d{1,3})(\.\d{1,3}){3}`)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	fd, err := os.OpenFile("/etc/hosts", os.O_RDWR|os.O_APPEND, 0666)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	defer fd.Close()
```
[the remark this answers is not in the document]
Comment (sunny-xiao) [reply pkg/server/chaosd/network.go:235]: help

[the remark this answers is not in the document]
Comment (WangXiangUSTC) [reply pkg/server/chaosd/network.go:235]: @muse-dev help

[the remark this answers is not in the document]
Comment (TomMD) [reply pkg/server/chaosd/network.go:235]: Sorry about that, we know people want to learn about the bug usually and have plans to help more there.

This error is about deferring a function that can return an error.  Close can error so deferring it swallows the error (potentially).

[the remark this answers is not in the document]
Comment (WangXiangUSTC) [reply pkg/server/chaosd/network.go:235]: Thanks, we got it

File: pkg/server/chaosd/network.go [hunk 2]
```diff
@@ -194,13 +209,75 @@ func (s *Server) applyTC(attack *core.NetworkCommand, ipset string, uid string)
...
+	}
+
+	fileBytes, err := ioutil.ReadFile("/etc/hosts.chaosd")
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	lines := strings.Split(string(fileBytes), "\n")
+
+	needle := "^(\\d{1,3})(\\.\\d{1,3}){3}.*\\b" + attack.DNSHost + "\\b.*"
```
Comment (WangXiangUSTC) [inline pkg/server/chaosd/network.go:225]: how about adding a comment to show an example string that can match this pattern?

File: pkg/server/chaosd/network.go [hunk 3]
```diff
@@ -38,6 +47,12 @@ func (networkAttack) Attack(options core.AttackConfig, env Environment) (err err
 
 	switch attack.Action {
 	case core.NetworkDNSAction:
+		if attack.NeedApplyEtcHosts() {
+			if err = env.Chaos.applyEtcHosts(attack, env.AttackUid); err != nil {
+				return errors.WithStack(err)
+			}
+		}
+
 		return env.Chaos.updateDNSServer(attack)
```
Comment (WangXiangUSTC) [inline pkg/server/chaosd/network.go:56]: how about adding a judgment for whether should update DNS server?

File: pkg/server/chaosd/network.go [hunk 4]
```diff
@@ -194,13 +209,75 @@ func (s *Server) applyTC(attack *core.NetworkCommand, ipset string, uid string)
 	return nil
 }
 
+func (s *Server) applyEtcHosts(attack *core.NetworkCommand, uid string) error {
+	backupCmd := exec.Command("/bin/bash", "-c", "mv /etc/hosts /etc/hosts.chaosd && touch /etc/hosts")
+	if err := backupCmd.Run(); err != nil {
```
Comment (WangXiangUSTC) [inline pkg/server/chaosd/network.go:214]: it is better to update to `backupCmd.Output`, so we can get some useful information if it runs failed

File: pkg/server/chaosd/network.go [hunk 5]
```diff
@@ -194,13 +209,75 @@ func (s *Server) applyTC(attack *core.NetworkCommand, ipset string, uid string)
...
+	}
+
+	lines := strings.Split(string(fileBytes), "\n")
+
+	needle := "^(\\d{1,3})(\\.\\d{1,3}){3}.*\\b" + attack.DNSHost + "\\b.*"
+	re, err := regexp.Compile(needle)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	reIp, err := regexp.Compile(`^(\d{1,3})(\.\d{1,3}){3}`)
```
Comment (WangXiangUSTC) [inline pkg/server/chaosd/network.go:230]: ditto

File: pkg/server/chaosd/network.go [hunk 6]
```diff
@@ -194,13 +213,106 @@ func (s *Server) applyTC(attack *core.NetworkCommand, ipset string, uid string)
 	return nil
 }
 
+func (s *Server) applyEtcHosts(attack *core.NetworkCommand, uid string, env Environment) error {
+	recoverFlag := true
+	cmd := "mv /etc/hosts /etc/hosts.chaosd." + uid + " && touch /etc/hosts"
+	backupCmd := exec.Command("/bin/bash", "-c", cmd)
```
[the remark this answers is not in the document]
Comment (sunny-xiao) [reply pkg/server/chaosd/network.go:219]: ignore

File: pkg/server/chaosd/network.go [hunk 7]
```diff
@@ -194,13 +213,106 @@ func (s *Server) applyTC(attack *core.NetworkCommand, ipset string, uid string)
...
+		}
+	}()
+
+	stdout, err := backupCmd.CombinedOutput()
+	if err != nil {
+		log.Error(backupCmd.String()+string(stdout), zap.Error(err))
+		return errors.WithStack(err)
+	}
+
+	fileBytes, err := ioutil.ReadFile("/etc/hosts.chaosd." + uid)
```
[the remark this answers is not in the document]
Comment (sunny-xiao) [reply pkg/server/chaosd/network.go:235]: ignore

File: pkg/server/chaosd/network.go [hunk 8]
```diff
@@ -194,13 +213,106 @@ func (s *Server) applyTC(attack *core.NetworkCommand, ipset string, uid string)
...
+	}
+	defer func() {
+		if err := fd.Close(); err != nil {
+			log.Error("Error closing file: %s\n", zap.Error(err))
+		}
+	}()
+
+	w := bufio.NewWriter(fd)
+
+	new_flag := true
```
Comment (WangXiangUSTC) [inline pkg/server/chaosd/network.go:270]: it's better to change the name to `newFlag`

**Review by WangXiangUSTC — COMMENTED** [review]
rest LGTM

**Comment by WangXiangUSTC** [comment]
@Andrewmatilde PTAL

**Comment by Andrewmatilde** [comment]
LGTM

**Review by Andrewmatilde — APPROVED** [review]
LGTM

**Pull request merged**